src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -335,6 +341,11 @@ protected override void Dispose(bool disposing)
 
         private void OnDependenciesChanged(object sender, DependenciesChangedEventArgs e)
         {
+            lock(_lockDataSourceVersions)
+            {
+                _requestedTreeUpdateDataSources.Add(e.DataSourceVersions);
```

Comment:
I don't think this change will fix the issue correctly.  There are couple problems:
1, it looks like your _requestedTreeUpdateDataSources is just a small buffer to hold data source versions, and sort them.  This is not a correct fix. Your sub tree can take dependencies on different set of sources, and the data source version coming out doesn't have to match each other, you have to merge versions from different sources, instead of sorting them.
2, the buffer only holds the a small set of changes which is waiting to be processed, if one of your sub tree provider is slow, it might not send data to you, until you finish processing the rest, so your sorting won't work.
3, when there are multiple data sources, the sub tree might get changes in different orders and it can be on (1, 2,1), and another is on (2,1,1), you cannot sort them at all.
4, even more complex, some version numbers are depending on others, so the value can go down, when others goes up.  The primary reason is when you switch from a debug to retail, data source ports on the retail might be lower than the debug side.  This is a design issue on the CPS side, so we have to work together to resolve.  For now, you might ignore it for a while.  But you do have to fix the rest issues, or the change won't work.